### ISSUE TITLE
ConvAdapter code

### ISSUE THREAD

**Issue by meeselizabeth** [Hhhhhhao/Conv-Adapter#1]
Hi! 

When is the code becoming available? Really interested in your work!

Best

**Comment by Hhhhhhao** [comment]
Hi, we will release the code by the end of May. Sorry for the delay!

**Comment by sarkar0613** [comment]
I'm also looking forward to your project!Thanks

**Comment by Hhhhhhao** [comment]
Code is up. Please mainly refer to the model code since the code is very old, and I might not have time currently to make it up-to-date.

**Comment by meeselizabeth** [comment]
Hi, I tried adding ConvAdapter in a segresnet by placing it parallel to the convolutions in all the resblocks (in downlayers and uplayers) like: 
`class ResBlockAda(nn.Module):
   
    def __init__(
        self,
        spatial_dims: int,
        in_channels: int,
        norm: Union[Tuple, str],
        kernel_size: int = 3,
        act: Union[Tuple, str] = ("RELU", {"inplace": True}),
        adapter_size: int = 16
    ) -> None:

        super().__init__()

        if kernel_size % 2 != 1:
            raise AssertionError("kernel_size should be an odd number.")

        self.norm1 = get_norm_layer(name=norm, spatial_dims=spatial_dims, channels=in_channels)
        self.norm2 = get_norm_layer(name=norm, spatial_dims=spatial_dims, channels=in_channels)
        self.act = get_act_layer(act)
        self.conv1 = get_conv_layer(spatial_dims, in_channels=in_channels, out_channels=in_channels)
        self.conv2 = get_conv_layer(spatial_dims, in_channels=in_channels, out_channels=in_channels)

        self.conv_adapter = ConvAdapter(in_channels, in_channels, 
                                             kernel_size=3, 
                                             padding=1,
                                             width=in_channels // adapter_size, 
                                             stride=1, 
                                             groups=in_channels // adapter_size, 
                                             dilation=1,
                                             act_layer=nn.ReLU)
        self.tuning_method = 'conv_adapt'

    def forward(self, x):
        identity = x

        x_ada = self.conv_adapter(x)

        x = self.norm1(x)
        x = self.act(x)
        x = self.conv1(x)
        x = x + x_ada
        #x = self.conv1(x) + self.conv_adapter(x)

        x = self.norm2(x)
        x = self.act(x)
        x = self.conv2(x)
        x = x + x_ada
        #x = self.conv2(x) + self.conv_adapter(x)

        x += identity

        return x` 

My segresnet has 4.7M parameters, and when I add ConvAdapter, and only train the adapter's parameters while freezing the pretrained model, I am training 29589 (0.03M) parameters. Very nice to have only a few parameters trained, however the performance is absolutely terrible 😅. Am I doing something wrong here? Hope you can help me!

Best

**Comment by Hhhhhhao** [comment]
Maybe add x_ada only after self.conv2 instead of both conv1 and conv2? 
Also in general you need large pre-trained models to achieve reasonable performance with parameter-efficient adaptation.

**Comment by meeselizabeth** [comment]
Thank you, I'll try only after conv2! 

Yeah but I would want to make it applicable for smaller models, but yes that is quite a challenge and maybe impossible..
